### ISSUE TITLE
[macOS] Failed to notarize application

### ISSUE THREAD

**Issue by itlancer** [airsdk/Adobe-Runtime-Support#1015]
## Problem Description
Failed to notarize macOS bundled AIR application.
When I try to notarize AIR application for macOS packaged as .pkg then I got errors.

Tested with latest AIR SDK 33.1.1.554 and multiple AIR applications (even without native extensions) with macOS Big Sur.
The same errors in all cases.

Related issue: https://github.com/airsdk/Adobe-Runtime-Support/issues/254
But solution of it linked to external resource which not available anymore.

## Steps to Reproduce
1) Package any AIR application. Packaging command looks like:
`adt -package -storetype pkcs12 -keystore cert.p12 -storepass pass -keypass pass -target bundle -storetype KeychainStore -alias "Developer ID Application: ..." MyApp.app descriptor.xml MyApp.swf`
It uses new code signing option mentioned here: https://github.com/airsdk/Adobe-Runtime-Support/discussions/966#discussioncomment-1004647
and here: https://github.com/airsdk/Adobe-Runtime-Support/issues/889#issuecomment-885075553
2) Try to notarize your app. Command looks like:
`xcrun altool -t osx -f "MyApp.pkg" --primary-bundle-id MyAppID --notarize-app`

**Actual Result:**
Notarization failed with errors:
```
{
  "logFormatVersion": 1,
  "jobId": "28a304ca-7261-2e9b-c021-0f1bbg4c0689",
  "status": "Invalid",
  "statusSummary": "Archive contains critical validation errors",
  "statusCode": 4000,
  "archiveFilename": "MyApp.pkg",
  "uploadDate": "2021-07-27T00:34:09Z",
  "sha256": "1caf8eab86727805e5fc28c4749708ab1d6731235e67a104066c3e8264311a3a",
  "ticketContents": null,
  "issues": [
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/WebKit.dylib",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/WebKit.dylib",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/A2712Enabler",
      "message": "The binary is not signed.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/A2712Enabler",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/A2712Enabler",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/A2712Enabler",
      "message": "The binary is not signed with a valid Developer ID certificate.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/A2712Enabler",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/A2712Enabler",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/Flash MyApp.plugin/Contents/MacOS/FlashMyApp-10.6",
      "message": "The binary is not signed.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "MyApp.pkg/MyApp.pkg Contents/Payload/Applications/MyApp.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/Flash MyApp.plugin/Contents/MacOS/FlashMyApp-10.6",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "x86_64"
    }
  ]
}
```

**Expected Result:**
Application will be notarized successfully and could be used for distribution.

## Known Workarounds
This works: https://github.com/airsdk/Adobe-Runtime-Support/issues/1015#issuecomment-888041940

Also you can use notarizing script from here: https://github.com/tuarua/WebViewANE/tree/master/mac_installer
It remove "problem" files:
```
rm -r "$APP.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/Adobe AIR.vch"
rm -r "$APP.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/WebKit.dylib"
rm -r "$APP.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/Flash Player.plugin"
rm -r "$APP.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/WebKit"
rm -r "$APP.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/__MACOSX"
rm -r "$APP.app/Contents/Frameworks/Adobe AIR.framework/Versions/1.0/Resources/A2712Enabler"
```
And notarization will passed successfully.
But `HTMLoader` functionality will be broken. So its not workaround for us.

**Comment by ajwfrost** [comment]
It looks like the codesign process we used there has just failed silently... we can double-check on this and see if we can get some extra diagnostics..

**Comment by ajwfrost** [comment]
As a workaround: to sign the app properly you would need to do the below commands:
```
find "MyApp.app" -name "*.framework"  -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
find "MyApp.app" -name "*.plugin"     -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
find "MyApp.app" -name "*.dylib"      -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
find "MyApp.app" -name "A2712Enabler" -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v MyApp.app
```

and for the 'ent.plist' file:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
        <key>com.apple.security.cs.allow-dyld-environment-variables</key>
        <true/>
        <key>com.apple.security.cs.allow-jit</key>
        <true/>
        <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
        <true/>
        <key>com.apple.security.cs.disable-library-validation</key>
        <true/>
</dict>
</plist>
```
plus whatever additional entitlements you would need in there from your app descriptor file.

**Comment by ajwfrost** [comment]
... and just checking, when your using ADT, do you see any output? We should be echoing anything that comes from the codesign tool via stderr...

thanks

**Comment by itlancer** [comment]
@ajwfrost 
Your workaround did the trick! Thanks! It works fine. Mentioned as workaround in original post.
Hope it will be fixed properly in next AIR SDK versions.

There is nothing suspicious when ADT executing, just:
```
/Users/username/Desktop/MyApp/adt15373669833217758205.tmp/temp: replacing existing signature
/Users/username/Desktop/MyApp/adt15373669833217758205.tmp/temp: signed app bundle with Mach-O universal (x86_64 arm64) [my.app.id]
```

**Comment by itlancer** [comment]
Fixed with latest AIR 33.1.1.620.
Thanks!
